apis/core/lib/domain/csv-source/update.ts

```diff
@@ -27,29 +27,38 @@ export async function update({
 
   csvSource.name = changed.name
   if (csvSource.setDialect(changed.dialect) && csvSource.associatedMedia.identifierLiteral) {
-    try {
-      csvSource.pointer.deleteOut(schema.error)
-      csvSource.columns = []
-
-      const fileStream = await fileStorage.loadFile(csvSource.associatedMedia.identifierLiteral) as Readable
-      const head = await loadFileHeadString(fileStream, 500)
-      const { header, rows } = await parse(head, {
-        delimiter: csvSource.dialect.delimiter,
-        quote: csvSource.dialect.quoteChar,
-      })
-
-      const sampleCol = sampleValues(header, rows)
-
-      for (let index = 0; index < header.length; index++) {
-        const name = header[index]
-        const column = csvSource.appendColumn({ name })
-        column.samples = sampleCol[index]
-      }
-    } catch (err) {
-      error(err)
-      csvSource.pointer.addOut(schema.error, err.message)
-    }
+    await updateColumns(csvSource, fileStorage)
   }
 
   return csvSource.pointer
 }
+
+export async function updateColumns(csvSource: CsvSource, fileStorage: s3.FileStorage): Promise<void> {
+  try {
+    if (!csvSource.associatedMedia.identifierLiteral) {
+      throw new Error('Key to file missing')
+    }
+
+    csvSource.pointer.deleteOut(schema.error)
+    csvSource.columns = []
+
+    const fileStream = await fileStorage.loadFile(csvSource.associatedMedia.identifierLiteral) as Readable
+    const head = await loadFileHeadString(fileStream, 500)
+    const { header, rows } = await parse(head, {
+      bom: true,
+      delimiter: csvSource.dialect.delimiter,
+      quote: csvSource.dialect.quoteChar,
+    })
+
+    const sampleCol = sampleValues(header, rows)
+
+    for (let index = 0; index < header.length; index++) {
+      const name = header[index]
+      const column = csvSource.appendColumn({ name })
```

Comment:
Do you rely on the `CsvColumn` identifiers to be stable when called multiple times?

I think this is not a good idea. For one, if the column order changes you will get multiple `dtype.order` triples, and also quite weird. Looking at the code of `appendColumn`, it will be like this:

Before (2 columns):

```turtle
<source>
  csvw:column <source/A> ; <source/B> .

<source/A> dtype:order 0 .
<source/B> dtype:order 1 .
```

After (updated with a file like `X;A;B`)

```turtle
<source>
  csvw:column <source/A> ; <source/B> ; <source/X> .

<source/A> dtype:order 0, 3 .
<source/B> dtype:order 1, 3 .
<source/X> dtype:order 2 .
```

Please add a test for updating with a CSV with new columns and/or columns in changed order

Reply:
Not exactly, we actually delete all columns beforehand. Hence the order is not preserved. But I can add a unit test for this.

This code is existing from the dialect change function, I did a refactoring to avoid duplication of the same thing.

Reply:
Where are the columns deleted, I think I missed it?